trained_models/Log_2022-03-14_09-05-26/parameters.txt
# -----------------------------------#
# Parameters of the training session #
# -----------------------------------#

# Input parameters
# ****************

dataset = ShapeNetPart_Pole
network_model = segmentation
num_classes = 12
in_points_dim = 3
in_features_dim = 4
in_radius = 1.000
input_threads = 8

# Model parameters
# ****************

architecture = simple resnetb resnetb_strided resnetb resnetb_strided resnetb_deformable resnetb_deformable_strided resnetb_deformable resnetb_deformable_strided resnetb_deformable nearest_upsample unary nearest_upsample unary nearest_upsample unary nearest_upsample unary
num_layers = 5
first_features_dim = 64
use_batch_norm = 1
batch_norm_momentum = 0.980

segmentation_ratio = 1.000

# KPConv parameters
# *****************

first_subsampling_dl = 0.020
num_kernel_points = 15
density_parameter = 5.000
fixed_kernel_points = center
KP_extent = 1.000
KP_influence = linear
convolution_mode = sum
trainable_positions = 0

modulated = 0

# Training parameters
# *******************

learning_rate = 0.010000
momentum = 0.980000
lr_decay_epochs = 1:0.971628 2:0.971628 3:0.971628 4:0.971628 5:0.971628 6:0.971628 7:0.971628 8:0.971628 9:0.971628 10:0.971628 11:0.971628 12:0.971628 13:0.971628 14:0.971628 15:0.971628 16:0.971628 17:0.971628 18:0.971628 19:0.971628 20:0.971628 21:0.971628 22:0.971628 23:0.971628 24:0.971628 25:0.971628 26:0.971628 27:0.971628 28:0.971628 29:0.971628 30:0.971628 31:0.971628 32:0.971628 33:0.971628 34:0.971628 35:0.971628 36:0.971628 37:0.971628 38:0.971628 39:0.971628 40:0.971628 41:0.971628 42:0.971628 43:0.971628 44:0.971628 45:0.971628 46:0.971628 47:0.971628 48:0.971628 49:0.971628 50:0.971628 51:0.971628 52:0.971628 53:0.971628 54:0.971628 55:0.971628 56:0.971628 57:0.971628 58:0.971628 59:0.971628 60:0.971628 61:0.971628 62:0.971628 63:0.971628 64:0.971628 65:0.971628 66:0.971628 67:0.971628 68:0.971628 69:0.971628 70:0.971628 71:0.971628 72:0.971628 73:0.971628 74:0.971628 75:0.971628 76:0.971628 77:0.971628 78:0.971628 79:0.971628 80:0.971628 81:0.971628 82:0.971628 83:0.971628 84:0.971628 85:0.971628 86:0.971628 87:0.971628 88:0.971628 89:0.971628 90:0.971628 91:0.971628 92:0.971628 93:0.971628 94:0.971628 95:0.971628 96:0.971628 97:0.971628 98:0.971628 99:0.971628 100:0.971628 101:0.971628 102:0.971628 103:0.971628 104:0.971628 105:0.971628 106:0.971628 107:0.971628 108:0.971628 109:0.971628 110:0.971628 111:0.971628 112:0.971628 113:0.971628 114:0.971628 115:0.971628 116:0.971628 117:0.971628 118:0.971628 119:0.971628 120:0.971628 121:0.971628 122:0.971628 123:0.971628 124:0.971628 125:0.971628 126:0.971628 127:0.971628 128:0.971628 129:0.971628 130:0.971628 131:0.971628 132:0.971628 133:0.971628 134:0.971628 135:0.971628 136:0.971628 137:0.971628 138:0.971628 139:0.971628 140:0.971628 141:0.971628 142:0.971628 143:0.971628 144:0.971628 145:0.971628 146:0.971628 147:0.971628 148:0.971628 149:0.971628 150:0.971628 151:0.971628 152:0.971628 153:0.971628 154:0.971628 155:0.971628 156:0.971628 157:0.971628 158:0.971628 159:0.971628 160:0.971628 161:0.971628 162:0.971628 163:0.971628 164:0.971628 165:0.971628 166:0.971628 167:0.971628 168:0.971628 169:0.971628 170:0.971628 171:0.971628 172:0.971628 173:0.971628 174:0.971628 175:0.971628 176:0.971628 177:0.971628 178:0.971628 179:0.971628 180:0.971628 181:0.971628 182:0.971628 183:0.971628 184:0.971628 185:0.971628 186:0.971628 187:0.971628 188:0.971628 189:0.971628 190:0.971628 191:0.971628 192:0.971628 193:0.971628 194:0.971628 195:0.971628 196:0.971628 197:0.971628 198:0.971628 199:0.971628 200:0.971628 201:0.971628 202:0.971628 203:0.971628 204:0.971628 205:0.971628 206:0.971628 207:0.971628 208:0.971628 209:0.971628 210:0.971628 211:0.971628 212:0.971628 213:0.971628 214:0.971628 215:0.971628 216:0.971628 217:0.971628 218:0.971628 219:0.971628 220:0.971628 221:0.971628 222:0.971628 223:0.971628 224:0.971628 225:0.971628 226:0.971628 227:0.971628 228:0.971628 229:0.971628 230:0.971628 231:0.971628 232:0.971628 233:0.971628 234:0.971628 235:0.971628 236:0.971628 237:0.971628 238:0.971628 239:0.971628 240:0.971628 241:0.971628 242:0.971628 243:0.971628 244:0.971628 245:0.971628 246:0.971628 247:0.971628 248:0.971628 249:0.971628 250:0.971628 251:0.971628 252:0.971628 253:0.971628 254:0.971628 255:0.971628 256:0.971628 257:0.971628 258:0.971628 259:0.971628 260:0.971628 261:0.971628 262:0.971628 263:0.971628 264:0.971628 265:0.971628 266:0.971628 267:0.971628 268:0.971628 269:0.971628 270:0.971628 271:0.971628 272:0.971628 273:0.971628 274:0.971628 275:0.971628 276:0.971628 277:0.971628 278:0.971628 279:0.971628 280:0.971628 281:0.971628 282:0.971628 283:0.971628 284:0.971628 285:0.971628 286:0.971628 287:0.971628 288:0.971628 289:0.971628 290:0.971628 291:0.971628 292:0.971628 293:0.971628 294:0.971628 295:0.971628 296:0.971628 297:0.971628 298:0.971628 299:0.971628 300:0.971628 301:0.971628 302:0.971628 303:0.971628 304:0.971628 305:0.971628 306:0.971628 307:0.971628 308:0.971628 309:0.971628 310:0.971628 311:0.971628 312:0.971628 313:0.971628 314:0.971628 315:0.971628 316:0.971628 317:0.971628 318:0.971628 319:0.971628 320:0.971628 321:0.971628 322:0.971628 323:0.971628 324:0.971628 325:0.971628 326:0.971628 327:0.971628 328:0.971628 329:0.971628 330:0.971628 331:0.971628 332:0.971628 333:0.971628 334:0.971628 335:0.971628 336:0.971628 337:0.971628 338:0.971628 339:0.971628 340:0.971628 341:0.971628 342:0.971628 343:0.971628 344:0.971628 345:0.971628 346:0.971628 347:0.971628 348:0.971628 349:0.971628 350:0.971628 351:0.971628 352:0.971628 353:0.971628 354:0.971628 355:0.971628 356:0.971628 357:0.971628 358:0.971628 359:0.971628 360:0.971628 361:0.971628 362:0.971628 363:0.971628 364:0.971628 365:0.971628 366:0.971628 367:0.971628 368:0.971628 369:0.971628 370:0.971628 371:0.971628 372:0.971628 373:0.971628 374:0.971628 375:0.971628 376:0.971628 377:0.971628 378:0.971628 379:0.971628 380:0.971628 381:0.971628 382:0.971628 383:0.971628 384:0.971628 385:0.971628 386:0.971628 387:0.971628 388:0.971628 389:0.971628 390:0.971628 391:0.971628 392:0.971628 393:0.971628 394:0.971628 395:0.971628 396:0.971628 397:0.971628 398:0.971628 399:0.971628 400:0.971628 401:0.971628 402:0.971628 403:0.971628 404:0.971628 405:0.971628 406:0.971628 407:0.971628 408:0.971628 409:0.971628 410:0.971628 411:0.971628 412:0.971628 413:0.971628 414:0.971628 415:0.971628 416:0.971628 417:0.971628 418:0.971628 419:0.971628 420:0.971628 421:0.971628 422:0.971628 423:0.971628 424:0.971628 425:0.971628 426:0.971628 427:0.971628 428:0.971628 429:0.971628 430:0.971628 431:0.971628 432:0.971628 433:0.971628 434:0.971628 435:0.971628 436:0.971628 437:0.971628 438:0.971628 439:0.971628 440:0.971628 441:0.971628 442:0.971628 443:0.971628 444:0.971628 445:0.971628 446:0.971628 447:0.971628 448:0.971628 449:0.971628 450:0.971628 451:0.971628 452:0.971628 453:0.971628 454:0.971628 455:0.971628 456:0.971628 457:0.971628 458:0.971628 459:0.971628 460:0.971628 461:0.971628 462:0.971628 463:0.971628 464:0.971628 465:0.971628 466:0.971628 467:0.971628 468:0.971628 469:0.971628 470:0.971628 471:0.971628 472:0.971628 473:0.971628 474:0.971628 475:0.971628 476:0.971628 477:0.971628 478:0.971628 479:0.971628 480:0.971628 481:0.971628 482:0.971628 483:0.971628 484:0.971628 485:0.971628 486:0.971628 487:0.971628 488:0.971628 489:0.971628 490:0.971628 491:0.971628 492:0.971628 493:0.971628 494:0.971628 495:0.971628 496:0.971628 497:0.971628 498:0.971628 499:0.971628 500:0.971628 501:0.971628 502:0.971628 503:0.971628 504:0.971628 505:0.971628 506:0.971628 507:0.971628 508:0.971628 509:0.971628 510:0.971628 511:0.971628 512:0.971628 513:0.971628 514:0.971628 515:0.971628 516:0.971628 517:0.971628 518:0.971628 519:0.971628 520:0.971628 521:0.971628 522:0.971628 523:0.971628 524:0.971628 525:0.971628 526:0.971628 527:0.971628 528:0.971628 529:0.971628 530:0.971628 531:0.971628 532:0.971628 533:0.971628 534:0.971628 535:0.971628 536:0.971628 537:0.971628 538:0.971628 539:0.971628 540:0.971628 541:0.971628 542:0.971628 543:0.971628 544:0.971628 545:0.971628 546:0.971628 547:0.971628 548:0.971628 549:0.971628 550:0.971628 551:0.971628 552:0.971628 553:0.971628 554:0.971628 555:0.971628 556:0.971628 557:0.971628 558:0.971628 559:0.971628 560:0.971628 561:0.971628 562:0.971628 563:0.971628 564:0.971628 565:0.971628 566:0.971628 567:0.971628 568:0.971628 569:0.971628 570:0.971628 571:0.971628 572:0.971628 573:0.971628 574:0.971628 575:0.971628 576:0.971628 577:0.971628 578:0.971628 579:0.971628 580:0.971628 581:0.971628 582:0.971628 583:0.971628 584:0.971628 585:0.971628 586:0.971628 587:0.971628 588:0.971628 589:0.971628 590:0.971628 591:0.971628 592:0.971628 593:0.971628 594:0.971628 595:0.971628 596:0.971628 597:0.971628 598:0.971628 599:0.971628 600:0.971628 601:0.971628 602:0.971628 603:0.971628 604:0.971628 605:0.971628 606:0.971628 607:0.971628 608:0.971628 609:0.971628 610:0.971628 611:0.971628 612:0.971628 613:0.971628 614:0.971628 615:0.971628 616:0.971628 617:0.971628 618:0.971628 619:0.971628 620:0.971628 621:0.971628 622:0.971628 623:0.971628 624:0.971628 625:0.971628 626:0.971628 627:0.971628 628:0.971628 629:0.971628 630:0.971628 631:0.971628 632:0.971628 633:0.971628 634:0.971628 635:0.971628 636:0.971628 637:0.971628 638:0.971628 639:0.971628 640:0.971628 641:0.971628 642:0.971628 643:0.971628 644:0.971628 645:0.971628 646:0.971628 647:0.971628 648:0.971628 649:0.971628 650:0.971628 651:0.971628 652:0.971628 653:0.971628 654:0.971628 655:0.971628 656:0.971628 657:0.971628 658:0.971628 659:0.971628 660:0.971628 661:0.971628 662:0.971628 663:0.971628 664:0.971628 665:0.971628 666:0.971628 667:0.971628 668:0.971628 669:0.971628 670:0.971628 671:0.971628 672:0.971628 673:0.971628 674:0.971628 675:0.971628 676:0.971628 677:0.971628 678:0.971628 679:0.971628 680:0.971628 681:0.971628 682:0.971628 683:0.971628 684:0.971628 685:0.971628 686:0.971628 687:0.971628 688:0.971628 689:0.971628 690:0.971628 691:0.971628 692:0.971628 693:0.971628 694:0.971628 695:0.971628 696:0.971628 697:0.971628 698:0.971628 699:0.971628 700:0.971628 701:0.971628 702:0.971628 703:0.971628 704:0.971628 705:0.971628 706:0.971628 707:0.971628 708:0.971628 709:0.971628 710:0.971628 711:0.971628 712:0.971628 713:0.971628 714:0.971628 715:0.971628 716:0.971628 717:0.971628 718:0.971628 719:0.971628 720:0.971628 721:0.971628 722:0.971628 723:0.971628 724:0.971628 725:0.971628 726:0.971628 727:0.971628 728:0.971628 729:0.971628 730:0.971628 731:0.971628 732:0.971628 733:0.971628 734:0.971628 735:0.971628 736:0.971628 737:0.971628 738:0.971628 739:0.971628 740:0.971628 741:0.971628 742:0.971628 743:0.971628 744:0.971628 745:0.971628 746:0.971628 747:0.971628 748:0.971628 749:0.971628 750:0.971628 751:0.971628 752:0.971628 753:0.971628 754:0.971628 755:0.971628 756:0.971628 757:0.971628 758:0.971628 759:0.971628 760:0.971628 761:0.971628 762:0.971628 763:0.971628 764:0.971628 765:0.971628 766:0.971628 767:0.971628 768:0.971628 769:0.971628 770:0.971628 771:0.971628 772:0.971628 773:0.971628 774:0.971628 775:0.971628 776:0.971628 777:0.971628 778:0.971628 779:0.971628 780:0.971628 781:0.971628 782:0.971628 783:0.971628 784:0.971628 785:0.971628 786:0.971628 787:0.971628 788:0.971628 789:0.971628 790:0.971628 791:0.971628 792:0.971628 793:0.971628 794:0.971628 795:0.971628 796:0.971628 797:0.971628 798:0.971628 799:0.971628 800:0.971628 801:0.971628 802:0.971628 803:0.971628 804:0.971628 805:0.971628 806:0.971628 807:0.971628 808:0.971628 809:0.971628 810:0.971628 811:0.971628 812:0.971628 813:0.971628 814:0.971628 815:0.971628 816:0.971628 817:0.971628 818:0.971628 819:0.971628 820:0.971628 821:0.971628 822:0.971628 823:0.971628 824:0.971628 825:0.971628 826:0.971628 827:0.971628 828:0.971628 829:0.971628 830:0.971628 831:0.971628 832:0.971628 833:0.971628 834:0.971628 835:0.971628 836:0.971628 837:0.971628 838:0.971628 839:0.971628 840:0.971628 841:0.971628 842:0.971628 843:0.971628 844:0.971628 845:0.971628 846:0.971628 847:0.971628 848:0.971628 849:0.971628 850:0.971628 851:0.971628 852:0.971628 853:0.971628 854:0.971628 855:0.971628 856:0.971628 857:0.971628 858:0.971628 859:0.971628 860:0.971628 861:0.971628 862:0.971628 863:0.971628 864:0.971628 865:0.971628 866:0.971628 867:0.971628 868:0.971628 869:0.971628 870:0.971628 871:0.971628 872:0.971628 873:0.971628 874:0.971628 875:0.971628 876:0.971628 877:0.971628 878:0.971628 879:0.971628 880:0.971628 881:0.971628 882:0.971628 883:0.971628 884:0.971628 885:0.971628 886:0.971628 887:0.971628 888:0.971628 889:0.971628 890:0.971628 891:0.971628 892:0.971628 893:0.971628 894:0.971628 895:0.971628 896:0.971628 897:0.971628 898:0.971628 899:0.971628 900:0.971628 901:0.971628 902:0.971628 903:0.971628 904:0.971628 905:0.971628 906:0.971628 907:0.971628 908:0.971628 909:0.971628 910:0.971628 911:0.971628 912:0.971628 913:0.971628 914:0.971628 915:0.971628 916:0.971628 917:0.971628 918:0.971628 919:0.971628 920:0.971628 921:0.971628 922:0.971628 923:0.971628 924:0.971628 925:0.971628 926:0.971628 927:0.971628 928:0.971628 929:0.971628 930:0.971628 931:0.971628 932:0.971628 933:0.971628 934:0.971628 935:0.971628 936:0.971628 937:0.971628 938:0.971628 939:0.971628 940:0.971628 941:0.971628 942:0.971628 943:0.971628 944:0.971628 945:0.971628 946:0.971628 947:0.971628 948:0.971628 949:0.971628 950:0.971628 951:0.971628 952:0.971628 953:0.971628 954:0.971628 955:0.971628 956:0.971628 957:0.971628 958:0.971628 959:0.971628 960:0.971628 961:0.971628 962:0.971628 963:0.971628 964:0.971628 965:0.971628 966:0.971628 967:0.971628 968:0.971628 969:0.971628 970:0.971628 971:0.971628 972:0.971628 973:0.971628 974:0.971628 975:0.971628 976:0.971628 977:0.971628 978:0.971628 979:0.971628 980:0.971628 981:0.971628 982:0.971628 983:0.971628 984:0.971628 985:0.971628 986:0.971628 987:0.971628 988:0.971628 989:0.971628 990:0.971628 991:0.971628 992:0.971628 993:0.971628 994:0.971628 995:0.971628 996:0.971628 997:0.971628 998:0.971628 999:0.971628
grad_clip_norm = 100.000000

augment_symmetries = 0 0 0
augment_rotation = none
augment_noise = 0.001000
augment_occlusion = none
augment_occlusion_ratio = 0.200
augment_occlusion_num = 1
augment_scale_anisotropic = 1
augment_scale_min = 0.900
augment_scale_max = 1.100
augment_color = 0.700

weights_decay = 0.001000
gaussian_decay = 0.001000
batch_averaged_loss = 0
offsets_loss = fitting
offsets_decay = 0.100000
batch_num = 16
max_epoch = 1000
epoch_steps = 13
validation_size = 50
snapshot_gap = 50


KPConv/utils/config.py
#
#
#      0=================================0
#      |    Kernel Point Convolutions    |
#      0=================================0
#
#
# ----------------------------------------------------------------------------------------------------------------------
#
#      Configuration class
#
# ----------------------------------------------------------------------------------------------------------------------
#
#      Hugues THOMAS - 11/06/2018
#


from os.path import join


class Config:
    """
    Class containing the parameters you want to modify for this dataset
    """

    ##################
    # Input parameters
    ##################

    # Dataset name
    dataset = ''

    # Type of network model
    network_model = 'segmentation'

    # Number of classes in the dataset
    num_classes = 0

    # Dimension of input points
    in_points_dim = 3

    # determines used features (Boolean)
    color_info = False
    intensity_info = False

    # Dimension of input features
    if color_info and intensity_info:
        in_features_dim = 8
    elif color_info:
        in_features_dim = 7
    elif intensity_info:
        in_features_dim = 5
    else:
        in_features_dim = 4

    # Radius of the input sphere (ignored for models, only used for point clouds)
    in_radius = 1.000

    # Number of CPU threads for the input pipeline
    input_threads = 8

    ##################
    # Model parameters
    ##################

    # Architecture definition. List of blocks
    architecture = []

    # Dimension of the first feature maps
    first_features_dim = 64

    # Batch normalization parameters
    use_batch_norm = True
    batch_norm_momentum = 0.99

    # For segmentation models : ratio between the segmented area and the input area
    segmentation_ratio = 1.0

    ###################
    # KPConv parameters
    ###################

    # First size of grid used for subsampling
    first_subsampling_dl = 0.05

    # Radius of the kernels in the first layer (deprecated)
    first_kernel_radius = 0.1

    # Number of points in the kernels
    num_kernel_points = 15

    # density of neighbors in kernel range
    # For each layer, support points are subsampled on a grid with dl = kernel_radius / density_parameter
    density_parameter = 3.0

    # Kernel point influence radius
    KP_extent = 1.000

    # Influence function when d < KP_extent. ('constant', 'linear', 'gaussian') When d > KP_extent, always zero
    KP_influence = 'gaussian'

    # Behavior of convolutions in ('closest', 'sum')
    # Decide if you sum all kernel point influences, or if you only take the influence of the closest KP
    convolution_mode = 'closest'

    # Fixed points in the kernel : 'none', 'center' or 'verticals'
    fixed_kernel_points = 'center'

    # Can the network learn kernel dispositions (deprecated)
    trainable_positions = False

    # Use modulateion in deformable convolutions
    modulated = False

    #####################
    # Training parameters
    #####################

    # Network optimizer parameters (learning rate and momentum)
    learning_rate = 1e-4
    momentum = 0.9

    # Learning rate decays. Dictionary of all decay values with their epoch {epoch: decay}.
    lr_decays = {200: 0.2, 300: 0.2}

    # Gradient clipping value (negative means no clipping)
    grad_clip_norm = 100.0

    # Augmentation parameters
    augment_scale_anisotropic = True
    augment_scale_min = 0.9
    augment_scale_max = 1.1
    augment_symmetries = [False, False, False]
    augment_rotation = 'vertical'
    augment_noise = 0.005
    augment_occlusion = 'planar'
    augment_occlusion_ratio = 0.2
    augment_occlusion_num = 1
    augment_color = 0.7

    # Regularization loss importance
    weights_decay = 1e-3

    # Gaussian loss
    gaussian_decay = 1e-3

    # Type of output loss with regard to batches when segmentation
    batch_averaged_loss = False

    # Point loss DPRECATED
    points_loss = ''
    points_decay = 1e-2

    # Offset regularization loss
    offsets_loss = 'permissive'
    offsets_decay = 1e-2

    # Number of batch
    batch_num = 10

    # Maximal number of epochs
    max_epoch = 1000

    # Number of steps per epochs
    epoch_steps = 1000

    # Number of validation examples per epoch
    validation_size = 100

    # Number of epoch between each snapshot
    snapshot_gap = 50

    # Do we nee to save convergence
    saving = True
    saving_path = None

    def __init__(self):
        """
        Class Initialyser
        """

        # Number of layers
        self.num_layers = len([block for block in self.architecture if 'pool' in block or 'strided' in block]) + 1

    def load(self, path):

        filename = join(path, 'parameters.txt')
        with open(filename, 'r') as f:
            lines = f.readlines()

        # Class variable dictionary
        for line in lines:
            line_info = line.split()
            if len(line_info) > 1 and line_info[0] != '#':

                if line_info[2] == 'None':
                    setattr(self, line_info[0], None)

                elif line_info[0] == 'lr_decay_epochs':
                    self.lr_decays = {int(b.split(':')[0]): float(b.split(':')[1]) for b in line_info[2:]}

                elif line_info[0] == 'architecture':
                    self.architecture = [b for b in line_info[2:]]

                elif line_info[0] == 'augment_symmetries':
                    self.augment_symmetries = [bool(int(b)) for b in line_info[2:]]

                elif line_info[0] == 'num_classes':
                    if len(line_info) > 3:
                        self.num_classes = [int(c) for c in line_info[2:]]
                    else:
                        self.num_classes = int(line_info[2])

                else:

                    attr_type = type(getattr(self, line_info[0]))
                    if attr_type == bool:
                        setattr(self, line_info[0], attr_type(int(line_info[2])))
                    else:
                        setattr(self, line_info[0], attr_type(line_info[2]))

        self.saving = True
        self.saving_path = path
        self.__init__()

    def save(self, path):

        with open(join(path, 'parameters.txt'), "w") as text_file:

            text_file.write('# -----------------------------------#\n')
            text_file.write('# Parameters of the training session #\n')
            text_file.write('# -----------------------------------#\n\n')

            # Input parameters
            text_file.write('# Input parameters\n')
            text_file.write('# ****************\n\n')
            text_file.write('dataset = {:s}\n'.format(self.dataset))
            text_file.write('network_model = {:s}\n'.format(self.network_model))
            if type(self.num_classes) is list:
                text_file.write('num_classes =')
                for n in self.num_classes:
                    text_file.write(' {:d}'.format(n))
                text_file.write('\n')
            else:
                text_file.write('num_classes = {:d}\n'.format(self.num_classes))
            text_file.write('in_points_dim = {:d}\n'.format(self.in_points_dim))
            text_file.write('in_features_dim = {:d}\n'.format(self.in_features_dim))
            text_file.write('in_radius = {:.3f}\n'.format(self.in_radius))
            text_file.write('input_threads = {:d}\n\n'.format(self.input_threads))

            # Model parameters
            text_file.write('# Model parameters\n')
            text_file.write('# ****************\n\n')

            text_file.write('architecture =')
            for a in self.architecture:
                text_file.write(' {:s}'.format(a))
            text_file.write('\n')
            text_file.write('num_layers = {:d}\n'.format(self.num_layers))
            text_file.write('first_features_dim = {:d}\n'.format(self.first_features_dim))
            text_file.write('use_batch_norm = {:d}\n'.format(int(self.use_batch_norm)))
            text_file.write('batch_norm_momentum = {:.3f}\n\n'.format(self.batch_norm_momentum))
            text_file.write('segmentation_ratio = {:.3f}\n\n'.format(self.segmentation_ratio))

            # KPConv parameters
            text_file.write('# KPConv parameters\n')
            text_file.write('# *****************\n\n')

            text_file.write('first_subsampling_dl = {:.3f}\n'.format(self.first_subsampling_dl))
            text_file.write('num_kernel_points = {:d}\n'.format(self.num_kernel_points))
            text_file.write('density_parameter = {:.3f}\n'.format(self.density_parameter))
            text_file.write('fixed_kernel_points = {:s}\n'.format(self.fixed_kernel_points))
            text_file.write('KP_extent = {:.3f}\n'.format(self.KP_extent))
            text_file.write('KP_influence = {:s}\n'.format(self.KP_influence))
            text_file.write('convolution_mode = {:s}\n'.format(self.convolution_mode))
            text_file.write('trainable_positions = {:d}\n\n'.format(int(self.trainable_positions)))
            text_file.write('modulated = {:d}\n\n'.format(int(self.modulated)))

            # Training parameters
            text_file.write('# Training parameters\n')
            text_file.write('# *******************\n\n')

            text_file.write('learning_rate = {:f}\n'.format(self.learning_rate))
            text_file.write('momentum = {:f}\n'.format(self.momentum))
            text_file.write('lr_decay_epochs =')
            for e, d in self.lr_decays.items():
                text_file.write(' {:d}:{:f}'.format(e, d))
            text_file.write('\n')
            text_file.write('grad_clip_norm = {:f}\n\n'.format(self.grad_clip_norm))


            text_file.write('augment_symmetries =')
            for a in self.augment_symmetries:
                text_file.write(' {:d}'.format(int(a)))
            text_file.write('\n')
            text_file.write('augment_rotation = {:s}\n'.format(self.augment_rotation))
            text_file.write('augment_noise = {:f}\n'.format(self.augment_noise))
            text_file.write('augment_occlusion = {:s}\n'.format(self.augment_occlusion))
            text_file.write('augment_occlusion_ratio = {:.3f}\n'.format(self.augment_occlusion_ratio))
            text_file.write('augment_occlusion_num = {:d}\n'.format(self.augment_occlusion_num))
            text_file.write('augment_scale_anisotropic = {:d}\n'.format(int(self.augment_scale_anisotropic)))
            text_file.write('augment_scale_min = {:.3f}\n'.format(self.augment_scale_min))
            text_file.write('augment_scale_max = {:.3f}\n'.format(self.augment_scale_max))
            text_file.write('augment_color = {:.3f}\n\n'.format(self.augment_color))

            text_file.write('weights_decay = {:f}\n'.format(self.weights_decay))
            text_file.write('gaussian_decay = {:f}\n'.format(self.gaussian_decay))
            text_file.write('batch_averaged_loss = {:d}\n'.format(int(self.batch_averaged_loss)))
            text_file.write('offsets_loss = {:s}\n'.format(self.offsets_loss))
            text_file.write('offsets_decay = {:f}\n'.format(self.offsets_decay))
            text_file.write('batch_num = {:d}\n'.format(self.batch_num))
            text_file.write('max_epoch = {:d}\n'.format(self.max_epoch))
            if self.epoch_steps is None:
                text_file.write('epoch_steps = None\n')
            else:
                text_file.write('epoch_steps = {:d}\n'.format(self.epoch_steps))
            text_file.write('validation_size = {:d}\n'.format(self.validation_size))
            text_file.write('snapshot_gap = {:d}\n'.format(self.snapshot_gap))



